FirebaseMLModelDownloader/Sources/ModelDownloadTask.swift

```diff
@@ -0,0 +1,163 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseCore
+
+enum DownloadStatus {
+  case notStarted
+  case inProgress
+  case completed
+}
+
+/// Progress and completion handlers for a model download.
+class DownloadHandlers {
+  typealias ProgressHandler = (Float) -> Void
+  typealias Completion = (Result<CustomModel, DownloadError>) -> Void
+
+  var progressHandler: ProgressHandler?
+  var completion: Completion
+
+  init(progressHandler: ProgressHandler?, completion: @escaping Completion) {
+    self.progressHandler = progressHandler
+    self.completion = completion
+  }
+}
+
+/// Manager for model downloads.
+class ModelDownloadTask: NSObject {
+  let app: FirebaseApp
+  var modelInfo: ModelInfo
```

Comment:
This `path` property in `modelInfo` gets updated after model download, so this will probably have to stay as a `var`.

Reply:
Ah, right! A couple questions here:
- if it's intended to be modify only from the class then `private(set) var` may better express this intention
- do we really need to modify the instance variable? Or maybe we may create a local copy in the method where we modify the model info to avoid unexpected accident modifications of the ivar?

It's mostly a suggestion so up to you.

Reply:
Good catch on `private(set)`, I made the change. 

Ultimately, we need modelInfo to be fully filled out before we can write it to user defaults, and we can't get all the info to fill it out at once. I wonder if there's any other way to do this without modifying the property?